src/Utilities/CssDocument.php

```diff
@@ -0,0 +1,286 @@
+     *           or an empty string if not from an `@media` rule);
+     *         - "selectors" (the CSS selector(s), e.g., "*" or "h1, h2");
+     *         - "declarations" (the semicolon-separated CSS declarations for that/those selector(s),
+     *           e.g., "color: red; height: 4px;").
+     */
+    public function getStyleRulesData(array $allowedMediaTypes): array
+    {
+        $splitCss = $this->splitCssAndMediaQuery($allowedMediaTypes);
+
+        $ruleMatches = [];
+        foreach ($splitCss as $cssPart) {
+            // process each part for selectors and definitions
+            \preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mi', $cssPart['css'], $matches, PREG_SET_ORDER);
+
+            foreach ($matches as $cssRule) {
+                $ruleMatches[] = [
+                    'media' => $cssPart['media'],
+                    'selectors' => $cssRule[1],
+                    'declarations' => $cssRule[2],
+                ];
+            }
+        }
+
+        return $ruleMatches;
+    }
+
+    /**
+     * Renders at-rules from the parsed CSS that are valid and not conditional group rules (i.e. not rules such as
+     * `@media` which contain style rules whose data is returned by {@see getStyleRulesData}).
+     *
+     * @return string
+     */
+    public function renderAtRules(): string
+    {
+        return $this->atRules;
+    }
+
+    /**
+     * Removes comments from the supplied CSS.
+     *
+     * @param string $css
+     *
+     * @return string CSS with the comments removed
+     */
+    private function removeCssComments(string $css): string
+    {
+        return \preg_replace('%/\\*[^*]*+(?:\\*(?!/)[^*]*+)*+\\*/%', '', $css);
+    }
+
+    /**
+     * Extracts `@import` and `@charset` rules from the supplied CSS.  These rules must not be preceded by any other
+     * rules, or they will be ignored.  (From the CSS 2.1 specification: "CSS 2.1 user agents must ignore any '@import'
+     * rule that occurs inside a block or after any non-ignored statement other than an @charset or an @import rule."
+     * Note also that `@charset` is case sensitive whereas `@import` is not.)
+     *
+     * @param string $css CSS with comments removed
+     *
+     * @return array{0: string, 1: string}
+     *         The first element is the CSS with the valid `@import` and `@charset` rules removed.  The second element
+     *         contains a concatenation of the valid `@import` rules, each followed by whatever whitespace followed it
+     *         in the original CSS (so that either unminified or minified formatting is preserved); if there were no
+     *         `@import` rules, it will be an empty string.  The (valid) `@charset` rules are discarded.
+     */
+    private function extractImportAndCharsetRules(string $css): array
+    {
+        $possiblyModifiedCss = $css;
+        $importRules = '';
+
+        while (
+            \preg_match(
+                '/^\\s*+(@((?i)import(?-i)|charset)\\s[^;]++;\\s*+)/',
```

Comment:
Should we move this to a constant? (Maybe then the `preg_match` will also fit in one line.)

Reply:
Moved to a constant and formatted the `preg_match` as one line.